### PR TITLE
[VPlan] Use DomTreeUpdater to automatically update DT for vector loop.

### PR DESCRIPTION
**Pull request by fhahn**
Use DTU to queue DominatorTree updates directly when connecting basic blocks during VPlan execution.

This simplifies DT updates and will also automatically allow updating the DT for the VPlan-native path as additional benefit in a follow-up

### CODE REVIEW

**Review by rengolin — COMMENTED** [review]
I'm surprised by the simplicity of this patch. Not that it's wrong, just that it sounds simpler than the previous code that was meant to avoid complexity. :)

IIUC, instead of doing the full analysis at the end (in `updateDomintorTree`) you do it at the BB creation time (`BB::execute`) and wrapping time (`VP::execute`). That avoids the need for a whole update in the end, which is great.

In that sense, this looks really good to me, but I'm not an expert in DT updater, so I'll let others more comfortable in the area to approve.

**Comment by fhahn** [comment]
> I'm surprised by the simplicity of this patch. Not that it's wrong, just that it sounds simpler than the previous code that was meant to avoid complexity. :)
> 
> IIUC, instead of doing the full analysis at the end (in `updateDomintorTree`) you do it at the BB creation time (`BB::execute`) and wrapping time (`VP::execute`). That avoids the need for a whole update in the end, which is great.
> 

Yes, that's the nice thing about the updater, all you need is to tell it which changes to the CFG has been made, it will efficiently apply those updates to the dominator tree. It's heavily used across various parts of LLVM, including common helpers like `SplitBlockImpl`  (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Utils/BasicBlockUtils.cpp#L1056) and many others in that file and elsewhere

File: llvm/lib/Transforms/Vectorize/VPlan.cpp [hunk 1]
```diff
@@ -399,8 +400,8 @@ void VPTransformState::packScalarIntoVectorValue(VPValue *Def,
   set(Def, VectorValue, Instance.Part);
 }
 
-BasicBlock *
-VPBasicBlock::createEmptyBasicBlock(VPTransformState::CFGState &CFG) {
+BasicBlock *VPBasicBlock::createEmptyBasicBlock(VPTransformState::CFGState &CFG,
+                                                DomTreeUpdater &DTU) {
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.cpp:404]: Should DTU reside inside CFGState, now that updating DT is integrated with modifying the CFG?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 2]
```diff
@@ -410,6 +408,9 @@ struct VPTransformState {
 
   /// VPlan-based type analysis.
   VPTypeAnalysis TypeAnalysis;
+
+  /// Updater for the DominatorTree.
+  DomTreeUpdater DTU;
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.h:413]: nit: place DTU in DT's position (after LI), or inside CFG?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 3]
```diff
@@ -2946,7 +2947,8 @@ class VPBasicBlock : public VPBlockBase {
 private:
   /// Create an IR BasicBlock to hold the output instructions generated by this
   /// VPBasicBlock, and return it. Update the CFGState accordingly.
-  BasicBlock *createEmptyBasicBlock(VPTransformState::CFGState &CFG);
+  BasicBlock *createEmptyBasicBlock(VPTransformState::CFGState &CFG,
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.h:2950]: nit: if DTU is part of CFGState, then "Update the CFGState accordingly" would also imply updating DT, which this method now does.

File: llvm/lib/Transforms/Vectorize/VPlan.cpp [hunk 4]
```diff
@@ -829,6 +832,9 @@ void VPlan::execute(VPTransformState *State) {
   BasicBlock *VectorPreHeader = State->CFG.PrevBB;
   State->Builder.SetInsertPoint(VectorPreHeader->getTerminator());
 
+  State->DTU.applyUpdates(
+      {{DominatorTree::Delete, VectorPreHeader, State->CFG.ExitBB}});
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.cpp:836]: ExitBB is (still, the single) successor of PrevBB=VectorPreHeader, should they first be disconnected before updating DTU with their edge's deletion?

**Review by ayalz — COMMENTED** [review]
Nice cleanup!

Worth separating into two NFC's: one simplifying how legacy updates its DT, and another which teaches native to start update its DT?

Worth noting that parts of DT external to VPlan's scope are still updated directly (i.e., skeletal parts).

File: llvm/lib/Transforms/Vectorize/VPlan.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.cpp]: Moved, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.cpp (hunk 4 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.cpp:838]: Yes, updated to remove the edge before

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Now inside CFG

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 3 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Done, thanks!

**Review by fhahn — COMMENTED** [review]
> Worth separating into two NFC's: one simplifying how legacy updates its DT, and another which teaches native to start update its DT?

Updated to only pass DT to DTU in non-native path, which makes DTU updates a no-op in that case

File: llvm/test/Transforms/LoopVectorize/outer_loop_test1.ll [hunk 5]
```diff
@@ -13,7 +13,7 @@
 ;   }
 ; }
 ;
-; RUN: opt -S -passes=loop-vectorize -enable-vplan-native-path -verify-loop-info < %s | FileCheck %s
+; RUN: opt -S -passes=loop-vectorize -enable-vplan-native-path -verify-loop-info -verify-dom-info < %s | FileCheck %s
```
Comment (ayalz) [inline llvm/test/Transforms/LoopVectorize/outer_loop_test1.ll:16]: Should that be added once DT is updated for native?

File: llvm/lib/Transforms/Vectorize/VPlan.cpp [hunk 6]
```diff
@@ -891,13 +899,9 @@ void VPlan::execute(VPTransformState *State) {
...
-  // We do not attempt to preserve DT for outer loop vectorization currently.
-  if (!EnableVPlanNativePath) {
-    BasicBlock *VectorHeaderBB = State->CFG.VPBB2IRBB[Header];
-    State->DT->addNewBlock(VectorHeaderBB, VectorPreHeader);
-    updateDominatorTree(State->DT, VectorHeaderBB, VectorLatchBB,
-                        State->CFG.ExitBB);
-  }
+  State->CFG.DTU.flush();
+  assert(EnableVPlanNativePath || State->CFG.DTU.getDomTree().verify(
+                                      DominatorTree::VerificationLevel::Fast));
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.cpp:904]:
```suggestion
  // DT is currently updated for non-native path only.
  assert(EnableVPlanNativePath || State->CFG.DTU.getDomTree().verify(
                                      DominatorTree::VerificationLevel::Fast));
```
(retaining the comment)

**Review by ayalz — APPROVED** [review]
LGTM!
Couple of last final nits, plus worth updating the summary that native is left for follow-up.

File: llvm/test/Transforms/LoopVectorize/outer_loop_test1.ll (hunk 5 above)
Comment (fhahn) [reply llvm/test/Transforms/LoopVectorize/outer_loop_test1.ll:16]: Removed, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.cpp (hunk 6 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.cpp:904]: Added back, thanks!

**Pull request merged**